llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp

```diff
@@ -0,0 +1,244 @@
+//===---- SYCLPropagateJointMatrixUsage.cpp - SYCLPropagateJointMatrixUsage Pass
+//--===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Pass propagates optional kernel features metadata through a module call graph
+// for sycl_ext_oneapi_matrix extension
+//
+// The pass consists of three main steps:
+//
+// I. It builds Function -> string of joint matrix types and sizes values
+// mapping for usage in step II
+// II. Propagates all the values from step I. to the top of the call graph
+// III. Generates metadata with values of joint matrix types and sizes
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/SYCLLowerIR/SYCLPropagateJointMatrixUsage.h"
+
+#include "llvm/ADT/SmallString.h"
+#include "llvm/IR/InstIterator.h"
+#include "llvm/IR/IntrinsicInst.h"
+
+#include <set>
+
+using namespace llvm;
+
+namespace {
+
+/// Returns true if the function is a SPIRV or SYCL builtin, e.g.
+/// _Z28__spirv_GlobalInvocationId_xv
+/// NB! This function was copied from sycl-post-link/ModuleSplitter.cpp and the
+/// definition of entry point (i.e. implementation of the function) should be in
+/// sync between those two.
+bool isSpirvSyclBuiltin(StringRef FName) {
+  if (!FName.consume_front("_Z"))
+    return false;
+  // now skip the digits
+  FName = FName.drop_while([](char C) { return std::isdigit(C); });
+
+  return FName.startswith("__spirv_") || FName.startswith("__sycl_");
+}
+
+bool isEntryPoint(const Function &F) {
```

Comment:
We probably could have just done a bottom-up call graph scan by going through function uses to avoid copying all this code here and reason about what is entry point and what is not. However, I'm fine with the current approach as well.

If that would be necessary, we can refactor later once we get a bit more experience with this functionality.